### PR TITLE
Merage aot features

### PR DESCRIPTION
**Pull request by groenenboomj**

### CODE REVIEW

**Comment by xinyazhang** [comment]
The PR needs some major reversions in order to make it as the unified one b/w AOTrition

1. Use ALL UPPER CASE for all `tl.constexpr` input arguments;
2. `extra_tokens_n` is not `tl.constexpr`. `tl.constexpr` means an extra template argument and exponential growth of AOTriton kernels.
3. `bias_type` should not be a string. I don't know how to pass string as `tl.constexpr` now. I can get to know but let's not create extra troubles by comparing strings in GPU.
4. Do not check input values to enable/disable functionalities. Notably, do not use `if bias_ptr is not None:`, but `if ENABLE_BIAS` and add `ENABLE_BIAS` to the argument list;
5. Do not add underscore before the kernel entrance, e.g. `_attn_fwd` -> `attn_fwd`. For AOTriton these functions are public.

**Comment by groenenboomj** [comment]
> The PR needs some major reversions in order to make it as the unified one b/w AOTrition
> 
> 1. Use ALL UPPER CASE for all `tl.constexpr` input arguments;
> 2. `extra_tokens_n` is not `tl.constexpr`. `tl.constexpr` means an extra template argument and exponential growth of AOTriton kernels.
> 3. `bias_type` should not be a string. I don't know how to pass string as `tl.constexpr` now. I can get to know but let's not create extra troubles by comparing strings in GPU.
> 4. Do not check input values to enable/disable functionalities. Notably, do not use `if bias_ptr is not None:`, but `if ENABLE_BIAS` and add `ENABLE_BIAS` to the argument list;
> 5. Do not add underscore before the kernel entrance, e.g. `_attn_fwd` -> `attn_fwd`. For AOTriton these functions are public.
> 6. Please move anything that needs `torch` out of this file. It's creating circular dependencies again.

Most of these are reasonable.  Let's discuss at the sync and I can get most of this in with the potential split.

File: python/perf-kernels/flash-attention.py [hunk 1]
```diff
@@ -563,10 +688,13 @@ def forward(ctx, q, k, v, causal, bias, sm_scale, split_kernel=False):
         o = torch.empty_like(q, dtype=v.dtype)
         M = torch.empty((q.shape[0] * q.shape[1], q.shape[2]), device=q.device, dtype=torch.float32)
 
+        philox_seed = 0x1BF52
```
Comment (vgokhale) [inline python/perf-kernels/flash-attention.py:691]: I assume these are random constants chosen for reproducibility? Comments for these magic numbers would be helpful.

File: python/perf-kernels/flash-attention.py [hunk 2]
```diff
@@ -159,53 +210,52 @@ def _attn_fwd(
...
+        order=(0, 1)
     )
-    O_block_ptr = tl.make_block_ptr(
-        base=Out + qvk_offset,
-        shape=(N_CTX, BLOCK_DMODEL),
-        strides=(stride_om, stride_on),
-        offsets=(start_m * BLOCK_M, 0),
-        block_shape=(BLOCK_M, BLOCK_DMODEL),
-        order=(1, 0),
+    v_offset = off_hz * stride_vh
```
Comment (vgokhale) [inline python/perf-kernels/flash-attention.py:251]: kv always has the same shape. It would be preferable to keep this as a single term, so that it can consolidate the VGPRs when computing the pointers. Q is outside the loop but this is inside the loop so using as few VGPRs as possible is helpful.

File: python/perf-kernels/flash-attention.py [hunk 3]
```diff
@@ -216,7 +266,7 @@ def _attn_fwd(
         elif bias_type == "matrix":
             bias_ptr = tl.make_block_ptr(
                 base=bias + ((off_hz % H) * stride_bh),
-                shape=(N_CTX, N_CTX),
+                shape=(seqlen_k, seqlen_k),
```
Comment (vgokhale) [inline python/perf-kernels/flash-attention.py:269]: should this be seqlen_q x seqlen_k since the intermediate tensor is the produc of Q x K?

File: python/perf-kernels/flash-attention.py [hunk 4]
```diff
@@ -234,37 +284,58 @@ def _attn_fwd(
...
+                offsets=(start_m * BLOCK_M, 0),
+                block_shape=(BLOCK_M, BLOCK_N),
+                order=(1, 0)
+                )
+    else:
+        encoded_softmax_block_ptr = 0
     if STAGE & 1:
         # We don't currently support causal masking and padding.
         tl.static_assert((STAGE != 3) or not need_padding)
         # equal to N_CTX if N_CTX is already a multiple of block_M
```
Comment (vgokhale) [inline python/perf-kernels/flash-attention.py:309]: Update comment?

File: python/perf-kernels/flash-attention.py [hunk 5]
```diff
@@ -127,6 +164,19 @@ def _attn_fwd_inner(
         m_ij = tl.maximum(m_i, tl.max(qk, 1))
         qk = qk - m_ij[:, None]
         p = tl.math.exp2(qk)
+        # CAVEAT: Must update l_ij before applying dropout
+        l_ij = tl.sum(p, 1)
+        # Note about the conflicts of Flash attention algorithm and PyTorch's CUDA implementation
+        # PyTorch needs to return softmax(qk) (dropout mask encoded in sign bits)
+        # While Flash attention paper computer the dropout AFTER exp2(qk- m_ij)
```
Comment (vgokhale) [inline python/perf-kernels/flash-attention.py:171]: nit: typo. compute

File: python/perf-kernels/flash-attention.py (hunk 2 above)
Comment (xinyazhang) [reply python/perf-kernels/flash-attention.py:251]: One problem is having the same shape does not mean having the same strides.

File: python/perf-kernels/flash-attention.py (hunk 5 above)
Comment (vgokhale) [inline python/perf-kernels/flash-attention.py:171]: > While Flash attention paper computer the dropout AFTER exp2(qk- m_ij)

I'm a bit confused by this. The CUDA impl also applies dropout after exp2(qk-l_ij).

[Here](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/transformers/cuda/flash_attn/flash_fwd_kernel.h#L408) is where exp2(qk - l_ij) is computed.

And [here](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/transformers/cuda/flash_attn/flash_fwd_kernel.h#L430) is where dropout is applied, after.

File: python/perf-kernels/flash-attention.py [hunk 6]
```diff
@@ -127,6 +164,19 @@ def _attn_fwd_inner(
...
+        if ENABLE_DROPOUT:
+            philox_offset = batch_philox_offset + start_m * BLOCK_M * seqlen_k + start_n
+            keep = dropout_mask(philox_seed, philox_offset, dropout_p, BLOCK_M, BLOCK_N, seqlen_k)
+            if RETURN_ENCODED_SOFTMAX:
+                tl.store(encoded_softmax_block_ptr, tl.where(keep, p, -p).to(encoded_softmax_block_ptr.type.element_ty))
+            p = tl.where(keep, p, 0.0)
+        elif RETURN_ENCODED_SOFTMAX:
+            tl.store(encoded_softmax_block_ptr, p.to(encoded_softmax_block_ptr.type.element_ty))
```
Comment (vgokhale) [inline python/perf-kernels/flash-attention.py:179]: Does this need to be boundary_checked if seqlen_k or seqlen_q isn't a multiple of BLOCK_N/M?

File: python/perf-kernels/flash-attention.py [hunk 7]
```diff
@@ -127,6 +164,19 @@ def _attn_fwd_inner(
...
+        # Note about the conflicts of Flash attention algorithm and PyTorch's CUDA implementation
+        # PyTorch needs to return softmax(qk) (dropout mask encoded in sign bits)
+        # While Flash attention paper computer the dropout AFTER exp2(qk- m_ij)
+        if ENABLE_DROPOUT:
+            philox_offset = batch_philox_offset + start_m * BLOCK_M * seqlen_k + start_n
+            keep = dropout_mask(philox_seed, philox_offset, dropout_p, BLOCK_M, BLOCK_N, seqlen_k)
+            if RETURN_ENCODED_SOFTMAX:
+                tl.store(encoded_softmax_block_ptr, tl.where(keep, p, -p).to(encoded_softmax_block_ptr.type.element_ty))
+            p = tl.where(keep, p, 0.0)
+        elif RETURN_ENCODED_SOFTMAX:
```
Comment (vgokhale) [inline python/perf-kernels/flash-attention.py:178]: Return_softmax implies dropout but converse is not true. Can we instead have

```
if return_softmax:
  apply dropout
  store softmax
else if dropout:
  apply_dropout
```

File: python/perf-kernels/flash-attention.py [hunk 8]
```diff
@@ -543,7 +663,12 @@ def forward(ctx, q, k, v, causal, bias, sm_scale, split_kernel=False):
         num_warps = 8 if Lq == 128 else 4
         pre_load_v = False if Lq == 128 else True
         grid = (triton.cdiv(q.shape[2], BLOCK_M), q.shape[0] * q.shape[1], 1)
-        
+
+        if return_encoded_softmax:
```
Comment (vgokhale) [inline python/perf-kernels/flash-attention.py:667]: Out in the world, there is very little detail on what exactly this is. It would be good to add information on that here as comment, if we know.

File: python/perf-kernels/flash-attention.py [hunk 9]
```diff
@@ -543,7 +663,12 @@ def forward(ctx, q, k, v, causal, bias, sm_scale, split_kernel=False):
...
+            encoded_softmax = torch.zeros((q.shape[0], q.shape[1], q.shape[2], k.shape[2]), device=q.device, dtype=torch.float32)
```
Comment (vgokhale) [inline python/perf-kernels/flash-attention.py:668]: torch.empty is faster.

File: python/perf-kernels/flash-attention.py (hunk 6 above)
Comment (xinyazhang) [reply python/perf-kernels/flash-attention.py:179]: Yes, we'll add this later.
`encoded_softmax` was not well designed in this kernel since it is fundamentally a debugging feature.
Suppose unaligned store is supported by the GPU, we won't even have plan to pad this tensor.

File: python/perf-kernels/flash-attention.py [hunk 10]
```diff
@@ -26,53 +28,87 @@
...
         )
-    return bias.expand(batch, nheads, seqlen, seqlen), bias_type
+    return bias.expand(batch, nheads, seqlen_q, seqlen_k), bias_type
 
 @triton.jit
 def _attn_fwd_inner(
     acc, l_i, m_i, q,
     K_block_ptr, V_block_ptr,
     start_m,
+    seqlen_q,
```
Comment (vgokhale) [inline python/perf-kernels/flash-attention.py:77]: This doesn't seem like it'd be used since in here, everything is about k/v

File: python/perf-kernels/flash-attention.py (hunk 7 above)
Comment (xinyazhang) [reply python/perf-kernels/flash-attention.py:178]: Nope, `return_softmax` is orthogonal to `dropout`. User can request softmax results even if dropout is disabled.
For example, we probably want to take a look at softmax values when debugging https://github.com/pytorch/pytorch/issues/116176

File: python/perf-kernels/flash-attention.py (hunk 6 above)
Comment (vgokhale) [reply python/perf-kernels/flash-attention.py:179]: I'd add at least a tl.static_assert to check for this (more preferable), or a comment and a TODO.

Debugging issues here at a future data because we accidentally enabled encoded_softmax with a non-multiple of block size sequence length will be painful.

File: python/perf-kernels/flash-attention.py (hunk 7 above)
Comment (vgokhale) [reply python/perf-kernels/flash-attention.py:178]: This is not how CUDA does it:

https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/transformers/cuda/flash_attn/flash_fwd_kernel.h#L419

This also brings me back to my comment below on adding a comment as to why this feature is needed. IT is not clear.

File: python/perf-kernels/flash-attention.py (hunk 9 above)
Comment (xinyazhang) [reply python/perf-kernels/flash-attention.py:668]: Anything related to `encoded_softmax` is for debugging and validation, and filling zeros is intentional.
We need to make sure there were actual writes to `encoded_softmax`, and torch.empty nullified this purpose.

File: python/perf-kernels/flash-attention.py (hunk 2 above)
Comment (vgokhale) [reply python/perf-kernels/flash-attention.py:251]: That is a fair point.

File: python/perf-kernels/flash-attention.py (hunk 9 above)
Comment (vgokhale) [reply python/perf-kernels/flash-attention.py:668]: I cannot comprehend the response, because I do not understand what encoded_softmax is for. It would be helpful to address my comment above on adding information about its purpose.

For example, I'm not sure how filling zeros helps. p can be zeros as well. Or are we relying on the fact that p won't be zeros because for it to be 0, qk must be inf and that would indicate a problem? It is not clear.

File: python/perf-kernels/flash-attention.py (hunk 6 above)
Comment (groenenboomj) [reply python/perf-kernels/flash-attention.py:179]: Noted, I'll add an assert.

File: python/perf-kernels/flash-attention.py (hunk 5 above)
Comment (groenenboomj) [reply python/perf-kernels/flash-attention.py:171]: The mask we get out of this is much easier and simpler to use in testing.  The testing was already adjusted.  The CUDA implementation differs from the paper, which is what is referenced here.

File: python/perf-kernels/flash-attention.py (hunk 7 above)
Comment (groenenboomj) [reply python/perf-kernels/flash-attention.py:178]: It is necessary for SDPA testing and integration into PyTorch.  The behavior we have is correct as per unit tests but it does diverge from this example.  From our perspective we are making sure we have correct behavior but do not need to 100% match the CUTLASS reference.

File: python/perf-kernels/flash-attention.py (hunk 10 above)
Comment (groenenboomj) [reply python/perf-kernels/flash-attention.py:77]: Noted, will verify.

File: python/perf-kernels/flash-attention.py (hunk 7 above)
Comment (xinyazhang) [reply python/perf-kernels/flash-attention.py:178]: > This is not how CUDA does it:
> 
> https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/transformers/cuda/flash_attn/flash_fwd_kernel.h#L419
> 
> This also brings me back to my comment below on adding a comment as to why this feature is needed. IT is not clear.

For `encoded_softmax` we are not going to follow CUDA's behavior, and if I read that correctly the two choices are still orthogonal. `apply_dropout` will be a noop if `dropout_p` == 0 @groenenboomj

File: python/perf-kernels/flash-attention.py (hunk 9 above)
Comment (xinyazhang) [reply python/perf-kernels/flash-attention.py:668]: softmax is literally `softmax(Q'K)`. `encoded_softmax` set the sign bit according to the `dropout_mask`.
We need this information to validate the correctness of dropout operations.

filling zero is a way to validate the chuck of memory is updated. If using `torch.empty`, how can you tell the data was actually written by the kernel, or simply some random garbage filled by a previous user of this chunk of memory?

Again, this variable is for debugging. Arguing with performance is nonsense.

File: python/perf-kernels/flash-attention.py (hunk 7 above)
Comment (vgokhale) [reply python/perf-kernels/flash-attention.py:178]: If pytorch is okay not following CUDA, that works for me.

Also, no, I don't think they're not orthogonal according to the implementation:
https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/transformers/cuda/flash_attn/flash_api.cpp#L604

File: python/perf-kernels/flash-attention.py (hunk 5 above)
Comment (vgokhale) [reply python/perf-kernels/flash-attention.py:171]: I think what I'm getting at is that the mask is applied at the wrong location. It needs to be applied after doing the softmax, which would be after the exponential.

Notice that dropout sets some values to 0.0f, and others are scaled by 1/(1-dropout_p). You do this scaling outside the loop on the PV result, which is correct *only* if the dropout is applied post exponential. Otherwise you must scale by the exp(1/(1-dropout_p).

File: python/perf-kernels/flash-attention.py (hunk 9 above)
Comment (vgokhale) [reply python/perf-kernels/flash-attention.py:668]: It would be helpful if you can please take a step back and try to understand what I am trying to get at.

My original comment is borne out of a lack of understanding of this feature. I am not trying to argue about anything, except may be for adding a comment somewhere here on what this thing is. I do not think that is too much to ask for, and I do not appreciate that being called nonsense.

You did add that information here, and it does help me understand it now, but it could have been quicker and more peaceful to just add the damn comment in the code.

Comment (vgokhale) [reply python/perf-kernels/flash-attention.py:668]: Let me add more context to my requests. The Triton team is responsible for this kernel. If there are issues here further down, that is on us. For us to be able to fix those issues, we need to understand all features included herein, including their purpose. 

I think we've been pretty open to anything that helps or makes life easy for AOT w.r.t this kernel and codebase. So I don't think it is too much to ask for details on new features being added here that we've not seen before, nor is available in the literature.

File: python/perf-kernels/flash-attention.py [hunk 11]
```diff
@@ -273,27 +345,43 @@ def _attn_fwd(
...
             2, offs_m, offs_n,
-            N_CTX, pre_load_v,
+            PRE_LOAD_V,
+            False, seqlen_aligned,
+            bias_ptr,
+            ENABLE_DROPOUT,
+            RETURN_ENCODED_SOFTMAX
         )
     # epilogue
     # write back m
```
Comment (vgokhale) [inline python/perf-kernels/flash-attention.py:359]: Can you move this to  where m_ptrs are computed just below?

**Review by vgokhale — APPROVED** [review]
